### ISSUE TITLE
Select signals in a loop vs Select loops watching signals

### ISSUE THREAD

**Issue by DaneSlattery** [ivmarkov/ruwm#10]
Hi

I wanted to ask about an architectural choice that is made in a few places in the codebase and hopefully learn from the distinction.

I observe a few situations like this:
```rust

async fn run()
{
  loop {
   match select(signal_a.wait(), signal_b.wait()){
      Either::First(first) => // handle task a,
      Either::Second(second) => // handle task b
     }
  }
}

```

In my understanding, it is possible for `signal_a` and `signal_b` to both trigger around the same time, and because they are both `wait`ed, this `run` function would consume both signals. If I am correct, this can lead to missed notifications, since only one of them will be handled.

I believe there are two ways around this:
The first is to not consume the notification if it is triggered. In `channel-bridge` a notification is immediately acknowledged in `poll_wait` if it is triggered. One could implement a method to clear the notification, similar to an interrupt.

The second way is to poll each signal in their own loop, which is how I've handled it:

```rust

async fn run()
{
 match select(task_a(), task_b()){
  Either::First(first) => // task a finished for some reason,
  Either::Second(second) => // task b finished
 }
}

async fn task_a(){
 loop {
   signal_a.wait().await
   // handle task a
  // 
 } 
}


async fn task_b(){
 loop {
   signal_b.wait().await
   // handle task b
  // 
 } 
}
```

Is my concern about signals firing off at the same time valid?

**Comment by DaneSlattery** [comment]
Wrote this little test program, and it seems like both signals are caught, still wondering if this is something to be concerned about though :
```rust
use channel_bridge::notification::Notification;
use embassy_futures::{
    block_on,
    select::{select, Either},
};
fn main() {
    // It is necessary to call this function once. Otherwise some patches to the runtime
    // implemented by esp-idf-sys might not link properly. See https://github.com/esp-rs/esp-idf-template/issues/71
    esp_idf_svc::sys::link_patches();
    esp_idf_svc::timer::embassy_time_driver::link();

    // Bind the log crate to the ESP Logging facilities
    esp_idf_svc::log::EspLogger::initialize_default();
    signaler();
    block_on(run());
    log::info!("Hello, world!");
}

static SIGNAL_A: Notification = Notification::new();
static SIGNAL_B: Notification = Notification::new();

async fn run() {
    loop {
        match select(SIGNAL_A.wait(), SIGNAL_B.wait()).await {
            Either::First(_first) => {
                log::info!("Hello, signaled a");
                //
            }
            Either::Second(_second) => {
                log::info!("Hello, signaled b");
                //
            }
        }
    }
}

fn signaler() {
    SIGNAL_A.notify();
    SIGNAL_B.notify();
}
```

**Comment by ivmarkov** [comment]
> Is my concern about signals firing off at the same time valid?

TL;DR: No, they are not valid.

To elaborate, `select` operates in a single-threaded fashion, examining each of the two signals' futures _sequentially_. In other words, it calls `poll` on the future of the first signal (where the future is what you get from `async fn Signal::wait(&self) -> T` which is really a sugar for `fn Signal::wait(&self) -> impl Future<Output = T>`), and if the future returns `Poll::Ready<T>` which means **it pulls the value from the first signal and returns it to you**, then it does not even call `poll` on the second signal's future.

So on the next iteration of the loop, if the first signal is not ready yet, i.e. its future would return `Poll::Pending`, `select` will call `poll` the second signal's future and since it was ready already on the previous iteration of the loop, it will fetch its value and return it to you.

What you should be in general watching out for when using `embassy_future::select::select` (or any of the other `select` primitives in the Rust echosystem for that matter - including the `select!` macro monstrosity that I would suggest you avoid) is whether the future you are polling is "cancel-safe".

What I mean is the following: notice how after the call to `select` is over, BOTH of the signals' futures are dropped. Because they were created and then moved into the `select` function, so when this "function" exits ("function" in quotes, as this is not exactly a regular stackful function in the presence of "async") both futures are **dropped**.

Is dropping the futures are problem? The answer is: depends on the futures, unfortunately. Those which are created with `Signal::wait` are OK to be dropped at any time, because the information whether the signal had "triggered" (in your wording) is not kept inside the future (it just holds a `&ref` to the `Signal` object), but inside the `Signal` object **itself**. So when you drop the future and then you create another one, since they are stateless, nothing is lost.

Such futures are called "cancellation-safe".

Example of another cancellation-safe future:
- The future returned by `embedded_io_async::Read::read(buf: &mut [u8])` - If this future is dropped before it completes, nothing is lost - there won't be even a single byte read in your buffer. And if there were bytes ready for reading, you'll get them "next time" when you re-create the same future

Example of a future which is not cancellation-safe:
- The future returned by `embedded_io_async::Read::read_exact(buf: &mut [u8])` - this future calls `Read::read()` in a loop appending whatever bytes `read` returned to your buffer, until it gets full. In other words, it will only resolve when your buffer is full (and the input stream is over). What happens if you drop it before that? Dragons. As it might have read already _some_ bytes in your buffer, but because you dropped it before it completed, you'll **loose** those bytes. I mean, nothing will crash or anything like that (this is safe Rust after all) but your program will exhibit weird behavior if it is operating under the wrong assumptions. As in lose bytes from the input stream.

There is no silver bullet here. You just have to be careful w.r.t. cancellation safety and when you need it vs when you can tolerate cancellation-unsafe futures losing data when dropped.

There is a nice series of blogs on cancellation-safety by the `wg-async` group you might want to look after.
Also blogs from the Gods themselves: as in Boats and Niko, and a few others around them that have a deep treatment of the topic. Much better than my naive mumbling here.

**Comment by ivmarkov** [comment]
Closing because this is a general Rust `async` topic, not really related to `ruwm`. But feel free to further comment, if you feel like.

**Comment by DaneSlattery** [comment]
Thank you very much, fantastic insight as always. 
I think then that the only benefit of doing it the second way is that you can catch `signal_b` in `task_b()`'s loop if `task_a` is awaiting some other process.

**Comment by ivmarkov** [comment]
> Thank you very much, fantastic insight as always. I think then that the only benefit of doing it the second way is that you can catch `signal_b` in `task_b()`'s loop if `task_a` is awaiting some other process.

Both the "first" and the "second" way are completely fine, and the difference between them (perhaps, _this_ is what you are saying?) is that they have different _semantics_:
- With the first approach, you are waiting on BOTH signals, and as long as one of these signals is ready, you continue further down the loop
- With the second approach, you have two separate "tasks" (really, just big complex futures containing loops themselves) which are each waiting on one of these two signals and then continuing their own loops whenever their "own" signal completes.

What semantics you want depends on what your program is supposed to do. Which is an orthogonal topic to the "cancellation-safety" topic which was your original concern.

Or to spin it another way: most futures with a few exceptions (especially those from the `embassy` echosystem, `embassy-sync` in particular) are cancellation safe. You "larger" futures (as all your code is just futures composition) might or might not be cancellation-safe. Whether that's a problem depends on your concrete program / code snippet _semantics_ and nothing else.

If you are wondering about the difference between tasks and futures.... if you don't use an "executor" there is no difference really. What you call "task_a" and "task_b" are really just futures (or functions returning futures, that is). `select` and `join` by the way, are - amongst other things - a way to get by _without_ an executor and still exhibit concurrency (but not parallelism). Or to spin it another way - every executor is just a glorified, giant future (a composition of all futures you push down its throat as "tasks") which just has a lower latency when polling those futures. But much more restrictions as how these futures are composed. They are composed as if you do `select` or `join` on ALL of them.

Here's the perfect treatment from Boats on the topic (where he talks about "intra-task" concurrency, as a unique property of async Rust, as opposed to multi-threaded blocking Rust): https://without.boats/blog/let-futures-be-futures/

**Comment by DaneSlattery** [comment]
Just came back to this to say that there is a concern about the order of tasks in this fashion, regardless of cancellation safety. 

Let's say in my example that `SIGNAL_A` was triggering every 1s and the handler for `SIGNAL_A` was longer than 1s. It would drown out any notifcations on `SIGNAL_B`. You could re-order the tasks so that `SIGNAL_B` is checked first, or watch each signal in a separate loop.

**Comment by ivmarkov** [comment]
> Just came back to this to say that there is a concern about the order of tasks in this fashion, regardless of cancellation safety.
> 
> Let's say in my example that `SIGNAL_A` was triggering every 1s and the handler for `SIGNAL_A` was longer than 1s. It would drown out any notifcations on `SIGNAL_B`. You could re-order the tasks so that `SIGNAL_B` is checked first, or watch each signal in a separate loop.

That's true - `embassy_futures::select` does have a bias I think in that in case multiple futures are ready, it always gives priority to the earlier futures enumerated as its arguments, in that these are simply polled first.

If you want fairness, you have to use a "real" executor. Or perhaps look at [an alternative select impl?](https://github.com/rust-lang/futures-rs/issues/1805).

With that said, I think your use case is also a bit on the edge, because if your TASK_A task that is triggered by SIGNAL_A takes longer than 1s to execute, you anyway have a problem, as it will not meet its deadlines and will cause the other tasks scheduled in the same executor (TASK_B / SIGNAL_B) to starve. It is just that with an unbiased select or with an executor this starvation will not be constant, and TASK_B would at least get a chance to execute.

**Comment by ivmarkov** [comment]
> You could re-order the tasks so that SIGNAL_B is checked first, or watch each signal in a separate loop.

I think watching each signal in a separate loop won't help when the separate loops are ultimately combined again with `embassy_futures::select` or `join`. You are just moving the problem but not solving it.
